schema_salad/metaschema/salad.md

```diff
@@ -196,6 +197,7 @@ It is a fatal error if the document is not valid YAML.
 
 A Salad document must consist only of either a single root object or an
```

Comment:
What happens to `saladVersion` in case of an array of objects? Each object must provide its own `saladVersion`?

Reply:
At least, it is necessary to declare `saladVersion` at the root level.
Otherwise, we cannot decide whether to allow the newly introduced features such as nested typeDSL and map/union types.

I guess there are two options to declare `saladVersion` for the array of objects.
1. As @GlassOfWhiskey mentioned, each object in the array must declare `saladVersion`
1. I assume that the most root level documents use the explicit context. If my assumption is correct, it is reasonable to enforce `saladVersion` at the root level and to allow the imported documents inherit the `saladVersion` in the parent level.

Reply:
Option 1 is easier to be documented but it is more impractical for users.
Option 2 requires to distinguish between root Salad documents and imported Salad documents (those that are included through the `$import` statement), clearly define the two concepts, and state that a Salad document can be used as a root Salad document only if it specifies a `saladVersion` field. Conversely, if imported documents do not specify a `saladVersion` they inherit the one from their root Salad doument.

Another problem here is dealing with mixed versions, as in CWL. What if a Salad 2.0 document imports a Salad 3.0 document, and viceversa? Is this something that Salad allows?

Reply:
> Another problem here is dealing with mixed versions, as in CWL. What if a Salad 2.0 document imports a Salad 3.0 document, and viceversa? Is this something that Salad allows?

I prefer to take a [similar approach as CWL](https://www.commonwl.org/v1.2/Workflow.html#Syntax).
That is,
- Implementation allows a root document to import documents with newer and older `saladVersion`, and
- Each imported document is validated and processed as its declared `saladVersion` rather than its parent `saladVersion`.

---
Here is the corresponding description in CWL v1.2.

> Implementations should allow workflows to reference documents of both newer and older CWL versions (up to the highest version of CWL supported by that implementation). Where the runtime environment or runtime behavior has changed between versions, for that portion of the execution an implementation must provide runtime environment and behavior consistent with the document's declared version. An implementation must not expose a newer feature when executing a document that specifies an older version that does not include that feature.

Reply:
There is a technical issue for it.
The current code generator seems to hold `saladVersion` [only at the root level](https://github.com/common-workflow-language/schema_salad/blob/57e9f045dfdc40a4078b074ccb555fe11931c178/schema_salad/codegen.py#L63) rather than schema object level.

To fix this issue, we may need to introduce `saladVersion` for each schema object in addition to the root explicit context.

Reply:
@tom-tan Is this fixed?

Reply:
Not yet fixed because there is a corner case to be discussed as shown below:

> I guess there are two options to declare saladVersion for the array of objects.
>
> 1.  As @GlassOfWhiskey mentioned, each object in the array must declare saladVersion
> 1. I assume that the most root level documents use the explicit context. If my assumption is correct, it is reasonable to enforce saladVersion at the root level and to allow the imported documents inherit the saladVersion in the parent level.

@GlassOfWhiskey and I prefer the option 2 but it needs changes that may affect all the code generators.
There are several ways for option 2.
- Introduce `saladVersion` for each schema object as mentioned [above](https://github.com/common-workflow-language/schema_salad/pull/861#discussion_r1762668083), or
- Implement schemaSaladUpgrader such as [`cwl-normalizer`](https://github.com/common-workflow-language/cwl-utils?tab=readme-ov-file#normalize-a-cwl-document)
  - The behavior of `typeDSL` blocks this option. Related: #863 

I prefer to introduce a upgrader for schema salad after fixing the `typeDSL` issue.
What do you think?